src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobWorker.cs

```diff
@@ -0,0 +1,87 @@
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Reindex
+{
+    /// <summary>
+    /// The worker responsible for running the export job tasks.
+    /// </summary>
+    public class ReindexJobWorker
+    {
+        private readonly Func<IScoped<IFhirOperationDataStore>> _fhirOperationDataStoreFactory;
+        private readonly ReindexJobConfiguration _reindexJobConfiguration;
+        private readonly Func<IReindexJobTask> _reindexJobTaskFactory;
+        private readonly ILogger _logger;
+
+        public ReindexJobWorker(Func<IScoped<IFhirOperationDataStore>> fhirOperationDataStoreFactory, IOptions<ReindexJobConfiguration> reindexJobConfiguration, Func<IReindexJobTask> reindexJobTaskFactory, ILogger<ReindexJobWorker> logger)
+        {
+            EnsureArg.IsNotNull(fhirOperationDataStoreFactory, nameof(fhirOperationDataStoreFactory));
+            EnsureArg.IsNotNull(reindexJobConfiguration?.Value, nameof(reindexJobConfiguration));
+            EnsureArg.IsNotNull(reindexJobTaskFactory, nameof(reindexJobTaskFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _fhirOperationDataStoreFactory = fhirOperationDataStoreFactory;
+            _reindexJobConfiguration = reindexJobConfiguration.Value;
+            _reindexJobTaskFactory = reindexJobTaskFactory;
+            _logger = logger;
+        }
+
+        public async Task ExecuteAsync(CancellationToken cancellationToken)
+        {
+            var runningTasks = new List<Task>();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    // Remove all completed tasks.
+                    runningTasks.RemoveAll(task => task.IsCompleted);
+
+                    // Get list of available jobs.
+                    if (runningTasks.Count < _reindexJobConfiguration.MaximumNumberOfConcurrentJobsAllowed)
+                    {
+                        using (IScoped<IFhirOperationDataStore> store = _fhirOperationDataStoreFactory())
+                        {
+                            // Query reindex job records
+                            IReadOnlyCollection<ReindexJobWrapper> jobs = await store.Value.AcquireReindexJobsAsync(
+                                _reindexJobConfiguration.MaximumNumberOfConcurrentJobsAllowed,
+                                _reindexJobConfiguration.JobHeartbeatTimeoutThreshold,
+                                cancellationToken);
+
+                            foreach (ReindexJobWrapper job in jobs)
+                            {
+                                _logger.LogTrace($"Picked up job: {job.JobRecord.Id}.");
+
+                                runningTasks.Add(_reindexJobTaskFactory().ExecuteAsync(job.JobRecord, job.ETag, cancellationToken));
+                            }
+                        }
+                    }
+
+                    await Task.Delay(_reindexJobConfiguration.JobPollingFrequency, cancellationToken);
```

Comment:
Perhaps move this to a `finally` block

Reply:
moving to the finally block causes the TaskCancelled exception to be not caught.